### PR TITLE
Add class Cargo and Passeger train

### PR DESCRIPTION
**Pull request by vadimaganim**

### CODE REVIEW

File: part-4/train.rb [hunk 1]
```diff
@@ -0,0 +1,63 @@
+# frozen_string_literal: true
+
+class Train
+  attr_accessor :number, :type, :speed, :count_cars,
+   :current_station, :train_route
+
+  def initialize(number)
+    @number = number
+    @type = type
```
Comment (IvanNosov) [inline part-4/train.rb]: remove this

**Comment by yashka713** [comment]
@vadimaganim Were are your `end line`s?)

File: part-4/cargo_train.rb [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+class CargoTrain < Train
+  attr_accessor :type
+
+  def initialize(number)
+    @type = 'cargo'
+    super
+  end
+
+  def add_car(car)
+    if @speed >= 0 && car.type == @type
```
Comment (yashka713) [inline part-4/cargo_train.rb]: Should speed be `>=0`? Does it check, that train stop?

File: part-4/passenger_train.rb [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+
+class PassengerTrain < Train
+  attr_accessor :type
+
+  def initialize(number)
+    @type = 'passenger'
+    super
+  end
+
+  def add_car(car)
```
Comment (yashka713) [inline part-4/passenger_train.rb]: The same as `CargoTrain.add_car`
Does `PassengerTrain` have a `car` inside?

File: part-4/cargo_car.rb [hunk 4]
```diff
@@ -0,0 +1,7 @@
+class CarCargo
```
Comment (yashka713) [inline part-4/cargo_car.rb]: Your class name has to be the same as the filename.

File: part-4/passenger_car.rb [hunk 5]
```diff
@@ -0,0 +1,7 @@
+class CarPassenger
```
Comment (yashka713) [inline part-4/passenger_car.rb]: Your class name has to be the same as the filename.

**Comment by yashka713** [comment]
> При добавлении вагона к поезду, объект вагона должен сохраняться во внутреннем массиве поезда

I could not find the `Carriage` class and `carriages` inside `Train`.

File: part-4/cargo_car.rb [hunk 6]
```diff
@@ -0,0 +1,6 @@
+class CargoCar
+  attr_accessor :car_type
```
Comment (IvanNosov) [inline part-4/cargo_car.rb]: make `:car_type` only readable

File: part-4/cargo_train.rb [hunk 7]
```diff
@@ -0,0 +1,9 @@
+require_relative './train.rb'
+require_relative './cargo_car.rb'
```
Comment (IvanNosov) [inline part-4/cargo_train.rb]: `require_relative 'train'`

File: part-4/main.rb [hunk 8]
```diff
@@ -0,0 +1,119 @@
+require_relative './station.rb'
+require_relative './passenger_train.rb'
+require_relative './cargo_train.rb'
+require_relative './route.rb'
```
Comment (IvanNosov) [inline part-4/main.rb]: require_relative `route`

File: part-4/passenger_car.rb [hunk 9]
```diff
@@ -0,0 +1,6 @@
+class PassengerCar
+  attr_accessor :car_type
```
Comment (IvanNosov) [inline part-4/passenger_car.rb]: make type readable only

File: part-4/passenger_car.rb [hunk 10]
```diff
@@ -0,0 +1,6 @@
...
+  def initialize
+    @car_type = 'passenger'
```
Comment (IvanNosov) [inline part-4/passenger_car.rb]: change this to symbol

File: part-4/passenger_train.rb [hunk 11]
```diff
@@ -0,0 +1,9 @@
+require_relative './train.rb'
+require_relative './passenger_car.rb'
```
Comment (IvanNosov) [inline part-4/passenger_train.rb]: why do you need this here?

File: part-4/passenger_train.rb [hunk 12]
```diff
@@ -0,0 +1,9 @@
...
+
+class PassengerTrain < Train
+  def initialize(number)
+    @train_type = 'passenger'
```
Comment (IvanNosov) [inline part-4/passenger_train.rb]: make it symbol not a string

File: part-4/train.rb [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+      @cars << car
+    else
+      p 'Speed is too much or car have another type'
+    end
+  end
+
+  def change_speed(speed)
+    if speed.positive?
+      @speed += speed
+    elsif @speed + speed.negative?
```
Comment (IvanNosov) [inline part-4/train.rb]: what this actually does?

File: part-4/train.rb [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+      prev_station: route_prev_station.name,
+      next_station: route_next_station.name
+    }
+  end
+
+  def go_to_station(station)
+    if @train_route.stations.include?(station)
+      @current_station = station
+    else
+      p "You don't have that station in the route"
```
Comment (IvanNosov) [inline part-4/train.rb]: remove `p`
also use single quotes for string without interpolation

File: part-4/cargo_car.rb [hunk 15]
```diff
@@ -0,0 +1,6 @@
...
+  def initialize
+    @car_type = 'cargo'
```
Comment (IvanNosov) [inline part-4/cargo_car.rb]: make it symbol

File: part-4/station.rb [hunk 16]
```diff
@@ -0,0 +1,25 @@
+
```
Comment (IvanNosov) [inline part-4/station.rb]: delete this line

File: part-4/train.rb (hunk 13 above)
Comment (vadimaganim) [reply part-4/train.rb]: @IvanNosov you think me need delete this piece of code ? because i don't use him

Comment (IvanNosov) [reply part-4/train.rb]: no, I asked you what does this code do?
`@speed + speed.negative?` what do you check here?

Comment (vadimaganim) [reply part-4/train.rb]: @IvanNosov before this look like @speed + speed < 0 but now me need add brackets mb ?(@speed + speed).negative?

Comment (vadimaganim) [reply part-4/train.rb]: @IvanNosov 
def change_speed(speed)
    if @speed + speed < 0
      @speed == 0
    else
      @speed += speed
    end
  end

File: part-4/passenger_train.rb [hunk 17]
```diff
@@ -0,0 +1,8 @@
+require_relative './train.rb'
```
Comment (IvanNosov) [inline part-4/passenger_train.rb]: remove this, you don't use it here

File: part-4/station.rb [hunk 18]
```diff
@@ -0,0 +1,23 @@
...
+    @name = name
+    @trains = []
+  end
+
+  def add_train(train)
+    @trains << train
+  end
+
+  def delete_train(train_number)
+    @trains.delete_if { |train| train == train_number }
```
Comment (IvanNosov) [inline part-4/station.rb]: `train == train_number` what do you comapare here?

Comment (IvanNosov) [reply part-4/station.rb]: I believe this should be `train.number == train_number`

File: part-4/train.rb [hunk 19]
```diff
@@ -0,0 +1,72 @@
+# frozen_string_literal: true
+class Train
+  attr_accessor :number, :cars, :current_station, :train_route
+  attr_reader :speed, :train_type
+  def initialize(number)
+    @number = number
+    @train_type = train_type
+    @speed = 0
+    @train_route = nil
```
Comment (IvanNosov) [inline part-4/train.rb]: why do you set it to `nil`?

File: part-4/train.rb [hunk 20]
```diff
@@ -0,0 +1,72 @@
...
+    @current_station = nil
```
Comment (IvanNosov) [inline part-4/train.rb]: why do you set it to nil?

File: part-4/station.rb (hunk 18 above)
Comment (vadimaganim) [reply part-4/station.rb]: @IvanNosov I changed name of parameter, because i give obj like argument not name and comparing obj

**Pull request merged**